src/main/java/mezz/jei/transfer/TransferOperation.java

```diff
@@ -0,0 +1,25 @@
+package mezz.jei.transfer;
+
+import net.minecraft.network.FriendlyByteBuf;
+import net.minecraft.world.entity.player.Player;
+import net.minecraft.world.inventory.AbstractContainerMenu;
+import net.minecraft.world.inventory.Slot;
+
+/**
+ * Represents transferring an ItemStack from inventorySlot to craftingSlot.
+ */
+public record TransferOperation(Slot inventorySlot, Slot craftingSlot) {
+	public static TransferOperation readPacketData(FriendlyByteBuf buf, AbstractContainerMenu container) {
+		int inventorySlotIndex = buf.readVarInt();
+		int craftingSlotIndex = buf.readVarInt();
+
+		Slot inventorySlot = container.getSlot(inventorySlotIndex);
+		Slot craftingSlot = container.getSlot(craftingSlotIndex);
```

Comment:
Especially if the client sends back indices getSlot may return null, so might be worth sanitizing that the slots exist and it isn't just a modified JEI instance sending bad data to try and crash the server.

Reply:
JEI's server packet handlers are pretty safe, wrapped in a try/catch/log here:
https://github.com/mezz/JustEnoughItems/blob/2df668b5ac4a8473b9837ad2785d0f5a4fb845a6/src/main/java/mezz/jei/network/PacketHandler.java#L43-L50

Reply:
Ah okay